### PR TITLE
New Adapter: MeloZen

### PR DESCRIPTION
**Pull request by benben2001**
This pull request introduces a new bidder adapter called "MeloZen" for Prebid Server.
Hi @onkarvhanumante Documentation for this adapter can be found in [PR #5467](https://github.com/prebid/prebid.github.io/pull/5467).

### CODE REVIEW

File: static/bidder-info/melozen.yaml
Comment (onkarvhanumante) [inline static/bidder-info/melozen.yaml:3]: endpoint is reachable
```
% curl -i --location --request POST 'https://prebid.melozen.com/rtb/v2/bid?publisher_id='
HTTP/1.1 204 No Content
Server: nginx/1.14.1
Date: Thu, 01 Aug 2024 05:56:03 GMT
Content-Type: application/json
Connection: keep-alive
```

File: static/bidder-info/melozen.yaml
Comment (onkarvhanumante) [inline static/bidder-info/melozen.yaml:8]: Prebid team has sent email to verify above maintainers email. Requesting to responding back on email thread

File: adapters/melozen/melozentest/exemplary/app-video.json
Comment (onkarvhanumante) [inline adapters/melozen/melozentest/exemplary/app-video.json:1]: as per `static/bidder-info/melozen.yaml` video media type is not supported for app.

should update bidder info yaml to support video media type

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:45]: nitpick: should be `// Extract Melozen Params`

File: static/bidder-info/melozen.yaml
Comment (onkarvhanumante) [reply static/bidder-info/melozen.yaml:8]: @benben2001, email bounced back. Requesting to confirm maintainer email

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:59]: `PubId` is mentioned as required ext param in bidder info yaml 

Prebid core framework will ensure that `PubId` is present before invoking `MakeRequests` function

Therefore this check can be removed

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:64]: @benben2001  we are discarding entire request if `PubId` is not present for any single imp.

you could collect this error and continue

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:156]: as per` static/bidder-info/melozen.yaml` video media type is not supported for app.

should update bidder info yaml to support video media type or remove `video` check

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:176]: as per` static/bidder-info/melozen.yaml` video media type is not supported for app.

should update bidder info yaml to support video media type or remove `video` check

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:122]: could use response utils here 

https://github.com/prebid/prebid-server/blob/dd07431ef54c38a2617c04e76b3e7bda9c21f569/adapters/appnexus/appnexus.go#L159-L165

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:61]: `buildEndpointURL` is simple function parsing endpoint template. This code can be inlined here

```
url, err := macros.ResolveMacros(a.endpointTemplate, macros.EndpointTemplateParams{PublisherID: strImpParams.PubId})

```

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:134]: `adapters.NewBidderResponse()`  initializes `Currency` as `USD`

therefore, no need to reinitialize `Currency` on line 134

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:72]: Code path is not covered. Should add json test for better coverage

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:121]: Code path is not covered. Should add json test for better coverage

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:155]: Code path is not covered. Should add json test for better coverage

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:160]: Code path is not covered. Should add json test for better coverage

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:200]: Code path is not covered. Should add json test for better coverage

**Review by onkarvhanumante — COMMENTED** [review]
@benben2001  requesting to address open comments

Also should link the Bidder docs PR in PR description

File: static/bidder-info/melozen.yaml
Comment (benben2001) [reply static/bidder-info/melozen.yaml:3]: Thank you for confirming that the endpoint is reachable and behaves as expected. I appreciate your help in ensuring our configurations are correctly set up!

File: static/bidder-info/melozen.yaml
Comment (benben2001) [reply static/bidder-info/melozen.yaml]: Thank you for pointing out the error with the email address. Indeed, there was a typo in the email provided. The correct email should be DSP@melodong.com. I have updated the static/bidder-info/melozen.yaml file with the correct information and pushed the changes to the PR. Please review the updated changes at your convenience. Thank you for your patience and assistance.

File: adapters/melozen/melozentest/exemplary/app-video.json
Comment (benben2001) [reply adapters/melozen/melozentest/exemplary/app-video.json:1]: Thank you for pointing out the omission of video media type support in the yaml configuration for both site and app. I have updated the bidder info in the yaml file to include video as a supported media type and have also added corresponding JSON tests (app-video.json and web-video.json) to ensure coverage. The PR has been updated accordingly. Please review the changes at your convenience. Thanks again for your guidance!

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for the feedback on the comment line. I've updated the comment to correctly reflect // Extract Melozen Params as suggested. The change is included in the latest commit.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for your suggestion. I've removed the check for PubId as it's already ensured by the Prebid core framework, as you pointed out. The code is now updated and simplified. The changes are included in the latest commit. Please review the updates when you have a moment.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for your feedback. I have revised the code to not exit immediately on errors when building the endpoint URL, but instead to log these errors and continue with processing other imps. This should enhance the robustness of our adapter's error handling. I appreciate your guidance and please let me know if there are any further improvements needed.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go:144]: Thank you for your feedback. I have updated the bidder-info/melozen.yaml file to reflect the correct media type support as suggested. The changes ensure that the video media type is appropriately handled. The updated configuration has been pushed to the PR. Please review the changes at your convenience. Thank you for guiding the improvements.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go:160]: Thank you for pointing out the need to align our code with the supported media types defined in bidder-info/melozen.yaml. I have updated the handling of media types within our adapter to ensure consistency with our declared capabilities. The changes have been made in the latest commit. Please review the updates and let me know if further adjustments are required.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for your suggestion to use response utilities for handling HTTP status codes. I have updated the code to incorporate response utils and added the corresponding JSON tests(response-status-400.json and response-status-not-200.json) to ensure everything functions correctly. The PR has been updated with these changes. Please review the updates at your convenience.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for the suggestion to inline the macros.ResolveMacros call. I agree that simplifying the code by removing the buildEndpointURL function makes the overall process more straightforward and reduces unnecessary abstraction. I have updated the code accordingly. Please let me know if there's anything else that needs attention!

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for pointing out the redundancy in setting the Currency after initializing the BidderResponse. I see now that it's already set to "USD" by default in NewBidderResponse(). I will remove the redundant line to clean up the code. Your detailed review is much appreciated!

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for pointing out the lack of test coverage for this code path. I have added a new JSON test file named response-status-not-200.json to ensure that scenarios with unexpected status codes are properly tested. This should enhance our test coverage and ensure the robustness of our error handling. The changes have been pushed to the PR for review.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for your feedback regarding the test coverage for the currency conversion logic. I have updated the code to collect errors during the conversion process and continue with the remaining tasks instead of returning immediately. This change enhances our error handling by allowing further processing even if an error occurs. I've also added appropriate tests to ensure comprehensive coverage of this behavior. The updates have been pushed to the PR for further review.

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for pointing out the test coverage issue. Upon further review, I realized that the code segment:
if impression.Audio != nil {
    impression.Audio = nil
}
is not necessary for our current logic. I've decided to remove it to streamline our codebase. This change will make the testing more focused and relevant to the necessary functionalities. I appreciate your careful review and helpful suggestions!

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go:184]: Thank you for highlighting the uncovered code path. I have added a new JSON test file, wrong-bid-ext.json, to cover the scenario you mentioned. This ensures our handling of malformed or missing bid.Ext is thoroughly tested. The update has been pushed to the PR for review.

**Comment by benben2001** [comment]
@onkarvhanumante Thank you for your guidance throughout the review process. I have addressed all open comments and made the necessary changes to the code as suggested. Additionally, I have updated the PR description to include a link to the Bidder documentation PR [here](https://github.com/prebid/prebid.github.io/pull/5467) for better context and completeness. Please let me know if there's anything else that needs to be revised or if further information is required.

File: static/bidder-info/melozen.yaml
Comment (gargcreation1992) [reply static/bidder-info/melozen.yaml]: Sent email again to the correct address.

Comment (benben2001) [reply static/bidder-info/melozen.yaml]: Hi @gargcreation1992 ,

I have sent the "received" confirmation to the specified email address as requested. Please let me know if there's anything else needed from my side.

Thank you for your assistance.

Best,
MeloZen

Comment (onkarvhanumante) [reply static/bidder-info/melozen.yaml]: received response from [DSP@melodong.com](mailto:DSP@melodong.com)

File: adapters/melozen/melozen.go
Comment (onkarvhanumante) [inline adapters/melozen/melozen.go:22]: replace `{bidder}` with `melozen`

**Review by onkarvhanumante — COMMENTED** [review]
@benben2001  just one comment https://github.com/prebid/prebid-server/pull/3784#discussion_r1708831184. Good to merge after addressing it

File: adapters/melozen/melozen.go
Comment (benben2001) [reply adapters/melozen/melozen.go]: Thank you for your helpful suggestion! I have updated the comment to specify "MeloZen" instead of "{bidder}" to clarify that this section is specifically designed for the MeloZen adapter. I appreciate your attention to detail and your patience as we refine the code. The updated comment has been pushed to the PR.

**Review by onkarvhanumante — APPROVED** [review]

**Review by onkarvhanumante — APPROVED** [review]

**Review by gargcreation1992 — APPROVED** [review]

**Comment by benben2001** [comment]
@gargcreation1992 @onkarvhanumante 
Hi Prebid Team,

Thank you very much for reviewing and merging my pull request for the MeloZen adapter. I really appreciate your time and effort in getting it integrated into the Prebid codebase.

I would like to kindly ask about the expected timeline for when this adapter will be included in the next release. Could you provide any insight into when that might happen?

If there’s anything I can do to help expedite the process, such as additional testing or documentation, please let me know. I’d be happy to assist in any way I can.

Thanks again for your support and for maintaining such a great project.

Best regards,
GitHub Handle

**Pull request merged**
